## Random Forest
![grape](https://user-images.githubusercontent.com/12748752/126882595-d1f5449e-14bb-4ab3-809c-292caf0858a1.png)
* We can create *`BaggingClassifier`* typically with *`max_samples`* set to the size of the training set and passing it a *`DecisionTreeClassifier`*. 
* But to use the ***`RandomForestClassifier`*** class, which is more convenient and optimized for `Decision Trees` (similarly, there is a ***`RandomForestRegressor`*** class for regression tasks)


#### Libraries, *`Make_ Moon`* Dtaset and *`train-test split`*

In [9]:
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from matplotlib.colors import ListedColormap
from matplotlib import pyplot as plt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_moons

X, y = make_moons(n_samples=500, noise=0.30, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [10]:
bag_clf = BaggingClassifier(
    DecisionTreeClassifier(splitter="random", max_leaf_nodes=16, random_state=42),
    n_estimators=500, max_samples=1.0, bootstrap=True, n_jobs=-1, random_state=42)

In [11]:
bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)

In [12]:
from sklearn.ensemble import RandomForestClassifier

rnd_clf = RandomForestClassifier(n_estimators=500, max_leaf_nodes=16, n_jobs=-1, random_state=42)
rnd_clf.fit(X_train, y_train)

y_pred_rf = rnd_clf.predict(X_test)

In [13]:
np.sum(y_pred == y_pred_rf) / len(y_pred)  # almost identical predictions

0.976

### *Extra-Trees* or _`Extremely Randomized Trees ensemble`_
![plum](https://user-images.githubusercontent.com/12748752/126882596-b9ba4645-7001-435e-9a3c-d4416a2543c1.png)
* Extra-Trees classifier using Scikit-Learn’s ExtraTreesClassifier class


### Feature Importance 
![plum](https://user-images.githubusercontent.com/12748752/126882596-b9ba4645-7001-435e-9a3c-d4416a2543c1.png)
* Yet another great quality of Random Forests is that they make it easy to measure the relative importance of each feature. 
* `Scikit-Learn` measures a feature’s importance by looking at **_how much the tree nodes that use that feature reduce impurity on average_** (across all trees in the forest).
* More precisely, it is a _weighted average_, where each node’s weight is equal to the number of training samples that are associated with it.
* `Scikit-Learn` computes this score automatically for each feature after training, then it scales the results so that the sum of all importances is equal to 1. 
* You can access the result using the ***feature_importances_*** variable.

* Lets see `RandomForestClassifier` on the `iris` dataset and outputs each feature’s importance.

In [14]:
from sklearn.datasets import load_iris
iris = load_iris()
rnd_clf = RandomForestClassifier(n_estimators=500, n_jobs=-1, random_state=42)
rnd_clf.fit(iris["data"], iris["target"])
for name, score in zip(iris["feature_names"], rnd_clf.feature_importances_):
    print(name, score)

sepal length (cm) 0.11249225099876375
sepal width (cm) 0.02311928828251033
petal length (cm) 0.4410304643639577
petal width (cm) 0.4233579963547682


In [15]:
rnd_clf.feature_importances_

array([0.11249225, 0.02311929, 0.44103046, 0.423358  ])

> * It seems that the most important features are the petal length (44%) and width (42%), 
> * while sepal length and width are rather unimportant in comparison (11% and 2%, respectively):